### PR TITLE
Update and Recalculate Fees after Order Adjusted in Backoffice

### LINKED ISSUE #12512: Update and recalculate fees button not working for payment fees after order adjusted in backoffice
## Description
The order total is not being updated with Payment fees that are % based after adding to an order in the backoffice, even after clicking the 'Update and recalculate fees' button.   

## Expected Behavior
Customer makes and order and choose a payment option with a % fee.  A transaction fee is added to the order. In the backend a product is added to the order, order total changes, 'update and recalcuate fees' button is clicked.  It would be expected that this would result in the transaction fee increasing to reflect the new order total.   

## Actual Behaviour
'Update and recalcuate fees' button is clicked, and the transaction fee is not changed. Thus the order total does not correctly reflect the payment fees that should be attached to the order.  

## Steps to Reproduce
1. Create a payment fee with a % fee.
2. Create a customer order selecting that payment fee.
3. Add a product to that order in the backend.
4. Click the 'Update and recalculate fees' button.
5. Note that the transaction fee doesn't not change to reflect the payment fees that should be charged on the increased order total.

## Animated Gif/Screenshot
Screenshot 2024-05-22 at 6 19 08 AM

https://github.com/openfoodfoundation/openfoodnetwork/assets/88801240/e13aeaaa-8483-41c6-8012-24e3541d4b11

## Workaround
Need to manually add an adjustment to the order to reflect the missing payment fees.  Many hub/shop managers may not notice the issue and thus fees would be getting charged incorrectly.  

## Severity
bug-s3: a feature is broken but there is a workaround

https://github.com/openfoodfoundation/openfoodnetwork/wiki/Bug-severity
-->

## Your Environment

* Version used: openfoodnetwork.ca
* Browser name and version: Firefox 126.0 (64-bit)
* Operating System and version (desktop or mobile): Mac 13.6.6 (22G630) desktop

### PR DESCRIPTION
**Pull request by cyrillefr**
#### What? Why?

- Closes #12512 

#### What should we test?
- log as an admin or hub's owner (classical/legacy mode)
- visit `admin/orders`
- pick an order from the list, one that is complete, but with a transaction fee associated to a payment method
    like a card payment
- one should see a non zero transaction fee under `ORDER ADJUSTMENTS`
- add a product from the `Select a variant` select box
- chose a quantity from the newly showed line
- click on the `+` sign to add the product/variant to the order
- the transaction fee should be updated

#### Extras
The `UPDATE AND RECALCULATE FEES` is for shipment fees only.
A click on it executes this line:
`@order.shipments.map(&:refresh_rates)` from the `edit` method in the `orders` controller.

But the recalculation is in the `ensure_correct_adjustment` method in the `payment` model.
That is triggered by an `after_save` hook.
That is triggered when (amongst others I guess) a new product is added in the dashboard.
Problem was, no recalculation of the amount was made in the update mode, only in the create mode.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [X] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: spec/system/admin/order_spec.rb [hunk 1]
```diff
@@ -172,6 +172,27 @@ def new_order_with_distribution(distributor, order_cycle)
...
+      login_as_admin
+      visit spree.edit_admin_order_path(order_with_fees)
+
+      transaction_fee = order_with_fees.all_adjustments.payment_fee.eligible.first.amount
+      expect(page.find("#order_adjustments").text).to have_content(transaction_fee)
+
+      select2_select product.name, from: 'add_variant_id', search: true
+      find('button.add_variant').click
+      sleep(1)
+
```
Comment (rioug) [inline spec/system/admin/order_spec.rb:188]: You should be able to remove the `sleep` by checking the variant is added on the page.

**Review by rioug — CHANGES_REQUESTED** [review]
Nice one, thanks @cyrillefr :pray: 
Could you try to remove the `sleep` the spec ? It's considered bad practice to have use `sleep`, they can be source of flackyness amongst other things.

**Review by rioug — APPROVED** [review]

**Review by mkllnk — COMMENTED** [review]
> The UPDATE AND RECALCULATE FEES is for shipment fees only.

Where does this spec come from?

I would have expected the fee to be updated when the recalculate-button is clicked, not when the product is added. The transaction fee is supposed to cover the cost of Stripe. If the customer paid already then the current fee is reflecting the cost incurred because of the finished transaction. The admin may add a product for free or the difference is paid in cash at delivery. That's my understanding but the instance managers would know better.

**Comment by cyrillefr** [comment]
Hello @mkllnk ,

This is what I understand from reading the source.
Clicking on `UPDATE AND RECALCULATE FEES` leads to the `edit` method in `orders_controller.rb`.

That is:
```ruby
def edit                                       
  @order.shipments.map(&:refresh_rates)
end
```
The `refresh_rates` method deals only with refreshing shipment rates.
Then, on save, are triggered these 2 methods:
```ruby
after_save :ensure_correct_adjustment, :update_adjustments
```
```ruby
def ensure_correct_adjustment
  if fee_adjustment
    fee_adjustment.originator = shipping_method
```
With `fee_adjustment`:
```ruby
def fee_adjustment
  @fee_adjustment ||= adjustments.shipping.first 
end
```
And `update_adjustments` leads to
```ruby
Spree::ItemAdjustments.new(self).update
```
So, there is no update regarding payment transaction.
Also, the button to recalculate is on the Shipment subpage.

**Comment by mkllnk** [comment]
> the button to recalculate is on the Shipment subpage.

I was talking about the button on the order edit page:

image

**Comment by cyrillefr** [comment]
Hello @mkllnk ,

Aargh I am sorry, I mis typed :), I am talking about the same button, the thing I wanted to highlight is: it is not in the `payments` subpage.

**Comment by cyrillefr** [comment]
Accidentally deleted my project branch. Restored and reopened.

**Comment by mkllnk** [comment]
@RachL, could you give some input here? Shall we add a recalculate button on the payments page?

**Comment by RachL** [comment]
@mkllnk I don't think it is a good idea to add a button in the payment page. Users won't see it and will forget to click on it. If this means we need to update the fees when we add the product and not when users are clicking on the update button of the edit page, then let's do it.

**Comment by mkllnk** [comment]
Okay, let's go ahead with this then.

**Review by mkllnk — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @cyrillefr ,

Thanks again for this contribution.

I've verified the issue: indeed, the value of a flat percent transaction fee was not updated, neither when a line item was added/removed, not when clicking the Update and Recalculate fees button.

After this PR, orders which were previously edited and still had an incorrect transaction fee, can now be corrected by clicking Update and Recalculate button:

Before clicking the button:

image

After clicking the button - we can see the fee was updated:

image

Also, adding a line item to the order automatically updates the value:

image

This looks great! Thank you, merging.

**Pull request merged**